### ISSUE TITLE
Load entire world data at the speed of lightning

### ISSUE THREAD

**Issue by crankywright** [felix-b/atc#238]
Currently, loading the default apt.dat takes ~30 seconds.
We will be loading even more data, listed below:

- X-Plane global apt.dat file in Custom Scenery/Global Airports - another 30 sec - should we load it?
- custom scenery airports.
- airport instrument procedures
- airspace geometry
- en-route ATC units

#### The problem

Most likely, Including the above bullets will make loading times unacceptable.
There are two problems with the current approach:

- text files (apt.dat) are parsed over and over again every time the sim is launched
- the loading process relies heavily on slow memory allocations (the runtime library heap)

#### Proposed solution

- perform loading process once and save the data to a cache, in a binary format optimized for fast loading
- every time the sim is launched, we check whether the cached data is still valid (it's invalid after AIRAC cycle update, custom scenery changes, or plugin version change).
- if the cache is valid, we load it fast
- if the cache is invalid, we repeat the loading process and save a new cache

#### Cache structure (tech C++ level)

How to build the cache:

- compose cache stream out of pre-allocated buffers ("cache buffers")
- use `placement new` to allocate individual objects within the cache buffers
- the objects must never allocate anything on the heap - only within the cache buffers
- since the objects are always added, and never removed, the deallocation operation is unnecessary 
- replace all pointers (currently `shared_ptr`) with a custom pointer (let's name it `buffer_ptr`), which:
  - stores offset in a buffer instead of address in memory
  - resolves object addresses within the buffers both during the build, and after the cache is loaded
  - implements the `->` operator, so no massive code rewrite is necessary

How to load the cache

- when the cache buffers (just large blobs of bytes) are loaded from the cache file, all the objects contained in them are ready for use without additional processing
- the only required operation is rebasing the `buffer_ptr` - storing starting address of every buffer in a variable, so that `buffer_ptr` can correctly resolve memory addresses.

**Comment by crankywright** [comment]
@Random-Ranger what do you think?

**Comment by Random-Ranger** [comment]
I agree with the fact that we need a way to cache datas in a binary way to speed up loading of the world's immutable objects.

Concerning the procedures, i don't think we should load all of them when the plugin is starting, we only need procedures for the airports involved in the running instance. Nevertheless, they could be cached in a binary way once loaded for the first time.

I am not a C++ guru so  please correct me if i'm  wrong.

I might be mistaking but the direct mapping of the live structure in a defined memory area, as i understand your proposal, seems like a lot of wrok to be done. As i see it, we would have to change the stl container to be aware of that memory mapping.

I think we could use google's protobuf library (used in libserver) to serialize and deserialize cachable objects on disk. Maybe this answer on stackoverflow [store-a-set-of-protobufs-on-disk](https://stackoverflow.com/questions/14227355/store-a-set-of-protobufs-on-disk) could be a starting point.

**Comment by felix-b** [comment]
@Random-Ranger 

My proposal indeed implies either:
- creating custom allocators for STL containers (https://blog.feabhas.com/2019/03/thanks-for-the-memory-allocator/). 
- if the above fails, creating our own containers. We don't need much, just `vector`, `unoredered_map`, and maybe `unordered_set`. 

As for storing serialized protobuf messages, it's definitely an alternative. I guess it will be significantly faster than parsing text files. With that, I guess it will be significantly slower than just loading ready-to-run blobs into memory, because every object will have to be transformed from protobuf message into a libworld object. We can make all libworld objects hold their state in a protobuf object to get rid of the transformation, but that would be a massive code rewrite, and also it will add coupling to protobuf, which I'd prefer to steer clear from.

An additional consideration to take into account, regardless of the approach we choose:

It can be that some of the objects will manage mutable state during the runtime (for example, `Frequency` will manage current transmissions - unless we find an elegant way to get rid of this "mutable runtime state" phenomenon). Such objects may need an additional "inflate" (or "hydrate") step when loaded, but I expect that to run very quickly.
